[master] Fix requisite performance issues and aggregation

bdrx312 opened the pull request:
### What does this PR do?
* add install of networkx
* fix aggregate to properly work with requisites
* fix requisite checking to not be exponential
* fix pkg aggregate to work when multiple states specify the same package; added some type hints to state.py and tagify in event.py to make the code easier to follow

### What issues does this PR fix or reference?
Fixes: #8210, #47154, #59123, #62439, #65304

### Previous Behavior
Applying states with many requisites took exponential time by the number of states with requisites.
Aggregating states did not honor requisites correctly.

### New Behavior
Applying many states with requisites is no longer exponential.
Aggregating states now correctly honors requisites.

### Merge requirements satisfied?

- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

s0undt3ch commented on salt/requisite.py:
```diff
@@ -0,0 +1,412 @@
+                          state=prereq_chunk["state"])
+        self.dag.add_edge(prereq_check_node, node_tag,
+                          RequisiteType.PREREQUIRED)
+        for u, _, key, data in self.dag.in_edges(
```
```suggestion
        for u, _, key, data in self.dag.in_edges(
```
What is `u` here?
Can we get a name that represents what this is?

bdrx312 replied:
1. This is common naming for usage of the graph library for the first node in the vertex `(u, v)`. I will trying to figure out a more domain specific name.

s0undt3ch replied:
> I will trying to figure out a more domain specific name.

Thanks

s0undt3ch commented on salt/requisite.py:
```diff
@@ -0,0 +1,412 @@
+
+        # Iterate over the nodes in topological order to get the correct
+        # ordering which takes requisites into account
+        for i, node in enumerate(
```
Why enumerate if you're not using `i`?

bdrx312 replied:
This went through several iterations and at one point I believe it was using the `i`. It looks like it can be cleaned up and removed now.

s0undt3ch commented on salt/requisite.py:
```diff
@@ -0,0 +1,412 @@
+
+            node_type = data['state']
+            added = False
+            for i, group in enumerate(groups_by_type[node_type]):
+                if len(group) >= max_group_size:
+                    continue
+                # Check if the node can be reached from any node in the group
+                first_node = next(iter(group))
+                agg_node = topo_order.get(first_node)
+                # Since we are iterating in topological order we know
+                # that there is no path from the current node to the
+                # node in the group; so we only need to check the path
+                # from the group node to the current node
+                reachable = nx.has_path(dag, agg_node or first_node, node)
+                if not reachable:
+                    # If not, add the node to the group
+                    if agg_node is None:
+                        # there is now more than one node for this
+                        # group so aggregate them
+                        agg_node = f'__aggregate_{node_type}_{i}__'
```
```suggestion
            for idx, group in enumerate(groups_by_type[node_type]):
                if len(group) >= max_group_size:
                    continue
                # Check if the node can be reached from any node in the group
                first_node = next(iter(group))
                agg_node = topo_order.get(first_node)
                # Since we are iterating in topological order we know
                # that there is no path from the current node to the
                # node in the group; so we only need to check the path
                # from the group node to the current node
                reachable = nx.has_path(dag, agg_node or first_node, node)
                if not reachable:
                    # If not, add the node to the group
                    if agg_node is None:
                        # there is now more than one node for this
                        # group so aggregate them
                        agg_node = f'__aggregate_{node_type}_{idx}__'
```

s0undt3ch commented on salt/state.py:
```diff
@@ -53,26 +57,24 @@
+
 log = logging.getLogger(__name__)
 
+# See https://docs.saltproject.io/en/latest/ref/states/layers.html for details on the naming
+# __exclude__ and __extend__ are list values; the other items are state items are
+# state items that have dict values. This could be cleaned up some by making
+# exclude and extend into dicts instead of lists so the type of all the values are
+# homogenous.
```
```suggestion
# See https://docs.saltproject.io/en/latest/ref/states/layers.html for details on the naming
# __exclude__ and __extend__ are list values; the other items are state items 
# that have dict values. This could be cleaned up some by making
# exclude and extend into dicts instead of lists so the type of all the values are
# homogeneous.
```

s0undt3ch commented on salt/utils/event.py:
```diff
@@ -185,17 +186,17 @@ def tagify(suffix="", prefix="", base=SALT):
         parts.append(suffix)
 
-    for index, _ in enumerate(parts):
+    for index, part in enumerate(parts):
```
```suggestion
    for part in parts:
```

s0undt3ch commented on tests/pytests/integration/test_requisites.py:
Why do you need a salt master and/or salt minion?
Also, you can't mock on integration tests.

These should either be unit or functional tests, see https://github.com/saltstack/salt/tree/master/tests/pytests/functional/modules/state/requisites

bdrx312 replied:
I could not find the documentation that describes what should go in functional versus what should go into integration. Since I am trying to test the top level state.py and not a state/execution module I was not sure where the tests should go. I can definitely use some guidance on the best way to write the tests for this.

s0undt3ch replied:
Rule of thumb:

* If you need to mock, those are likely unit tests
* If you don't need to mock, but you also do not need a salt master/minion running, those are functional tests (some mocking is possible here)
* If you need the master/minion combo running, those are integration tests.

bdrx312 replied:
I want to configure `state_aggregate: True` in the minion config and then run state.apply apply to test the sls. Here is one of the tests I was manually running:
```
cat > /etc/salt/minion <<'EOF'
file_client: local
master_type: disable
state_aggregate: True
EOF

mkdir /srv/salt /srv/pillar /srv/salt/_grains
cat > /srv/salt/top.sls <<'EOF'
base:
  '*':
    - test
EOF

cat > /srv/salt/testing.sls <<'EOF'
"first packages":
  pkg.installed:
    - pkgs:
      - hello

"listen to second packages":
  test.succeed_with_changes:
    - listen:
      - "second packages"

"second packages":
  pkg:
    - installed
    - pkgs:
      - cowsay
      - fortune-mod
      # - sl
      # - figlet
    - require:
      - "requirement"

"third packages":
  pkg.installed:
    - name: cowsay
    - require:
      - "test": "requirement"
EOF
```
https://docs.saltproject.io/en/latest/topics/tutorials/writing_tests.html says integration tests are the preferred method of testing, so I attempted to put it into integration and run a full state apply so that it tests the requisites and mod_aggregate. However, actually installing the packages was not necessary and would slow down the test and cause them not to be repeatable. Scanning through some of the test code I noticed there are several ways to call an sls file. What is the newest/recommended way to run the set `state_aggregate: True` and run sls files I have in test_requisites.py?

lkubb replied:
(Was summoned here from Slack)

The requisite system is far off from what I develop on usually, but I have a bit of experience with the Salt test suite, which is why I'm trying to help out here. You're working on an important contribution imho.

After having a look at the integration tests, some initial reactions:

* Running this as a sensible unit test might be hard.
* You don't need a master or minion to run while testing, so in theory they should be functional tests. I believe you can override the effective opts per test module/file by returning the overrides as a dict in a module-scoped fixture named `minion_config_overrides`.
* I don't think you need to use the `pkg` state module for testing aggregation behavior, instead you could whip up a nop state module that has a mod_aggregate function and reference it in your testing states, thus obviating the need for mocking pkg.

Synthesizing the previous points, I think something along the following lines would work (untested, but should be a solid base):

```python
import textwrap

import pytest


@pytest.fixture(scope="module")
def minion_config_overrides():
    return {
        "file_client": "local",
        "master_type": "disable",
        "state_aggregate": True,
    }


@pytest.fixture(scope="module", autouse=True)
def nop_aggregate_mod(loaders, state_tree):
    mod_contents = textwrap.dedent("""
        __virtualname__ = "aggr"


        def __virtual__():
            return __virtualname__()


        def test(name, aggrs=None, **kwargs):
            return {
                "name": name,
                "result": True,
                "comment": "",
                "changes": {
                    "aggrs": aggrs or [name]
                },
            }


        def mod_aggregate(low, chunks, running):
            # modeled after the pkg state module
            aggrs = []
            for chunk in chunks:
                tag = __utils__["state.gen_tag"](chunk)
                if tag in running:
                    # Already ran
                    continue
                if chunk.get("state") == "aggr":
                    if "__agg__" in chunk:
                        continue
                    # Check for the same function
                    if chunk.get("fun") != low.get("fun"):
                        continue

                    if "aggrs" in chunk:
                        aggrs.extend(chunk["aggrs"])
                        chunk["__agg__"] = True
                    elif "name" in chunk:
                        aggrs.append(chunk["name"])
                        chunk["__agg__"] = True
            if aggrs:
                if "aggrs" in low:
                    low["aggrs"].extend(aggrs)
                else:
                    low["aggrs"] = aggrs
            return low
    """)

    with pytest.helpers.temp_file("aggrs.py", state_tree / "_states", mod_contents):
        res = loaders.modules.saltutil.sync_all()
        assert "states" in res
        assert "aggr" in res["states"]
        loaders.reload_all()
        assert hasattr(loaders.states, "aggr")
        yield
    loaders.modules.saltutil.sync_all()
    loaders.reload_all()


def test_aggregate_requisites(state_tree, modules):
    """Test to ensure that aggregated states honor requisites"""
    sls_name = "requisite_aggregate_test"
    sls_contents = """
    "packages 1":
      aggr.test:
        - aggrs:
          - hello
    "listen to packages 2":
      test.succeed_with_changes:
        - listen:
          - "packages 2"
    "packages 2":
      aggr:
        - test
        - aggrs:
          - cowsay
          - fortune-mod
        - require:
          - "requirement"
    "packages 3":
      aggr.test:
        - name: cowsay
        - require:
          - "test": "requirement"
    "requirement":
      test.nop:
        - name: "requirement_name"
    """
    sls_tempfile = pytest.helpers.temp_file(f"{sls_name}.sls", state_tree, sls_contents)
    with sls_tempfile:
        # Apply the state file
        ret = modules.state.apply(sls_name)

         # etc... The assertions will change slightly.
```

s0undt3ch reviewed (dismissed):
Since you are adding a hard dependency, also add it [here](https://github.com/saltstack/salt/blob/6694a7a273d00567a4973b069096ec71017db63c/.pylintrc#L674)

s0undt3ch commented:
Please don't make huge commit messages.
Give it a good enough title, then feel free to expand and/or explain what was done.

s0undt3ch commented:
![image](https://github.com/saltstack/salt/assets/300048/bcd115c2-27db-434a-9f52-b6e42740a22a)

bdrx312 commented on salt/requisite.py:
This is a helper module for the state module for dealing with requisites, so it should belong in utils?

bdrx312 commented on salt/requisite.py:
```diff
@@ -0,0 +1,412 @@
+log = logging.getLogger(__name__)
+
+# See https://docs.saltproject.io/en/latest/ref/states/layers.html for details on the naming
+LowChunk = dict[str, Any]
```
This works in python 3.8 if I add `from __future__ import annotations`?

s0undt3ch replied:
> This works in python 3.8 if I add `from __future__ import annotations`?

You need the future import and the right syntax. It should be `:` not `=`.

bdrx312 replied:
I did some more research on this and could not find the documentation about this. All the documentation I could find shows that you can use `TypeAlias` `from typing` or in python 3.12 you can use the new `type` keyword. I did not use `TypeAlias`  because it was not added until python 3.10 and salt needs to support python 3.8. Can you provide some links to documentation on this?

s0undt3ch replied:
https://docs.python.org/3.8/library/__future__.html which links you to https://peps.python.org/pep-0563/

Wrong syntax and no futures import.

```
❯ python3.8 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 3, in <module>
    LowChunk = dict[str, Any]
TypeError: 'type' object is not subscriptable
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ python3.9 /tmp/foo.py
dict[str, typing.Any]
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)
```
```
❯ cat /tmp/foo.py
from typing import Any

LowChunk = dict[str, Any]

print(LowChunk)
```

Right syntax, you are defining a type, not a variable, no futures import.
```
❯ python3.8 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 3, in <module>
    LowChunk: dict[str, Any]
TypeError: 'type' object is not subscriptable
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ python3.9 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 5, in <module>
    print(LowChunk)
NameError: name 'LowChunk' is not defined
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
The above is actually good, because we defined the type, not a variable.
```
❯ cat /tmp/foo.py
from typing import Any

LowChunk: dict[str, Any]

print(LowChunk)
```

Right syntax, with the futures import:
```
❯ python3.8 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 7, in <module>
    print(LowChunk)
NameError: name 'LowChunk' is not defined
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ python3.9 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 7, in <module>
    print(LowChunk)
NameError: name 'LowChunk' is not defined
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ cat /tmp/foo.py
from __future__ import annotations

from typing import Any

LowChunk: dict[str, Any]

print(LowChunk)
```

bdrx312 replied:
Is there something I need to do to make type checkers like pyright and mypy work with this syntax?

bdrx312 commented:
> Please don't make huge commit messages. Give it a good enough title, then feel free to expand and/or explain what was done.

Can you elaborate on where you are suggesting to expand and/or explain what was was done? Do you mean break the message up into separate lines and have the first line of the commit message be short or is there a better place for describing the details of what was changed?

s0undt3ch commented:
> > Please don't make huge commit messages. Give it a good enough title, then feel free to expand and/or explain what was done.
> 
> Can you elaborate on where you are suggesting to expand and/or explain what was was done? Do you mean break the message up into separate lines and have the first line of the commit message be short or is there a better place for describing the details of what was changed?

This first like should be a short summary.
Then, two line breaks and you can go wild explaining what was done.

lkubb commented on salt/utils/requisite.py:
```diff
@@ -0,0 +1,457 @@
+
+class RequisiteType(str, Enum):
+    """Types of direct requisites"""
+
```
This crashes on Python 3.9, which is supported by Salt (https://github.com/python/mypy/issues/7591#issuecomment-1376636258 / https://github.com/astral-sh/ruff/issues/2346). It might be the cause of the SSH test failures on RHEL derivatives.

bdrx312 replied:
Ah thanks for the info, I was not aware of this issue on python 3.9.

bdrx312 replied:
Updated to use staticmethod_hack. I still need to figure out a solution to the importlib_metadata issue though.

lkubb replied:
I don't see anything wrong with proper one I proposed below, but I'm also not sure why it's not done like that currently.

You could check if including it fixes the rest of the integration tests and/or breaks anything else.

bdrx312 replied:
Thanks for you help.

lkubb replied:
You're welcome, hope this can land in 3008. (Thanks for creating this btw, I'm sure it was a major effort!)

Since you went pretty deep into the state compiler - if you have the capacity at some point - could you comment on https://github.com/saltstack/salt/issues/59959#issuecomment-2053688186? I would love to fix that issue (and possibly https://github.com/saltstack/salt/issues/49301) in the next release as well (based on this branch since its current state is just unwieldy).

lkubb commented:
On Fedora 38 with Python 3.11 as default, while running the SSH tests, I'm seeing the following crash of the thin client (not the master one), which seems to cause the `KeyError` on `sys.doc`:

```
Traceback (most recent call last):
  File "/var/tmp/.3a1413/pyall/salt/loader/lazy.py", line 770, in _load_module
    spec.loader.exec_module(mod)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/var/tmp/.3a1413/pyall/salt/modules/saltutil.py", line 22, in <module>
    import salt.client.ssh.client
  File "/var/tmp/.3a1413/pyall/salt/client/ssh/__init__.py", line 35, in <module>
    import salt.state
  File "/var/tmp/.3a1413/pyall/salt/state.py", line 31, in <module>
    import networkx as nx
  File "/var/tmp/.3a1413/py3/networkx/__init__.py", line 20, in <module>
    from networkx.utils.backends import _dispatch
  File "/var/tmp/.3a1413/py3/networkx/utils/backends.py", line 132, in <module>
    backends = _get_backends("networkx.plugins")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/tmp/.3a1413/py3/networkx/utils/backends.py", line 106, in _get_backends
    items = entry_points(group=group)
            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/importlib/metadata/__init__.py", line 1040, in entry_points
    return SelectableGroups.load(eps).select(**params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/importlib/metadata/__init__.py", line 476, in load
    ordered = sorted(eps, key=by_group)
              ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/importlib/metadata/__init__.py", line 1037, in <genexpr>
    eps = itertools.chain.from_iterable(
                                       ^
  File "/usr/lib64/python3.11/importlib/metadata/_itertools.py", line 16, in unique_everseen
    k = key(element)
        ^^^^^^^^^^^^
AttributeError: 'PathDistribution' object has no attribute '_normalized_name'
```

Ensuring the shim chooses Python 3.10 resolves the issue and makes the tests pass.

bdrx312 commented:
> On Fedora 38 with Python 3.11 as default, while running the SSH tests, I'm seeing the following crash of the thin client (not the master one), which seems to cause the `KeyError` on `sys.doc`:
> 
> ```
> Traceback (most recent call last):
>   File "/var/tmp/.3a1413/pyall/salt/loader/lazy.py", line 770, in _load_module
>     spec.loader.exec_module(mod)
>   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "/var/tmp/.3a1413/pyall/salt/modules/saltutil.py", line 22, in <module>
>     import salt.client.ssh.client
>   File "/var/tmp/.3a1413/pyall/salt/client/ssh/__init__.py", line 35, in <module>
>     import salt.state
>   File "/var/tmp/.3a1413/pyall/salt/state.py", line 31, in <module>
>     import networkx as nx
>   File "/var/tmp/.3a1413/py3/networkx/__init__.py", line 20, in <module>
>     from networkx.utils.backends import _dispatch
>   File "/var/tmp/.3a1413/py3/networkx/utils/backends.py", line 132, in <module>
>     backends = _get_backends("networkx.plugins")
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/var/tmp/.3a1413/py3/networkx/utils/backends.py", line 106, in _get_backends
>     items = entry_points(group=group)
>             ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/importlib/metadata/__init__.py", line 1040, in entry_points
>     return SelectableGroups.load(eps).select(**params)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/importlib/metadata/__init__.py", line 476, in load
>     ordered = sorted(eps, key=by_group)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/importlib/metadata/__init__.py", line 1037, in <genexpr>
>     eps = itertools.chain.from_iterable(
>                                        ^
>   File "/usr/lib64/python3.11/importlib/metadata/_itertools.py", line 16, in unique_everseen
>     k = key(element)
>         ^^^^^^^^^^^^
> AttributeError: 'PathDistribution' object has no attribute '_normalized_name'
> ```
> 
> Ensuring the shim chooses Python 3.10 resolves the issue and makes the tests pass, so not sure what causes this exactly, but it's probably related to the polyfill Salt does in `salt/_compat.py`/`salt/ext/importlib_metadata/`.

I found that it appears to be a bad interaction with `salt/ext/importlib_metadata` and `importlib.metadata`

lkubb commented:
> I also found that it appears to be a bad interaction with `salt/ext/importlib_metadata` and `importlib.metadata`

Yup, I think it overrides the stdlib `importlib.metadata`:

https://github.com/saltstack/salt/blob/aad71fdbcfcd7845b11e67f33a1173a22324ce8e/salt/ext/importlib_metadata/_compat.py#L40-L50

> importlib.metadata is available in all the versions of python supported by salt

Afaict the minimum required version of `importlib_metadata` is 3.3, which is only found in Py3.10+ per https://github.com/python/importlib_metadata?tab=readme-ov-file#compatibility, while Salt currently supports Py3.8+.

Is there a reason why Salt opts to not use the stdlib `importlib.metadata` on Py3.10+? Maybe @s0undt3ch has an idea how to proceed here.

Side note: I hate myself for it, but patching out the call before importing `networkx` makes this issue go away (we don't rely on networkx backends I assume).

bdrx312 committed:
Fix issues with requisites and aggregate

add install of networkx
fix aggregate to properly work with requisites
fix requisite checking to not be exponential
fix pkg aggregate to work when multiple states specify the same package
add some type hints to state.py to make the code easier to follow
fix case of pkg aggregate duplicate package.

bdrx312 commented:
Is there any idea or status on what release or date this might be targeted for?

lkubb commented:
@bdrx312 I took quite some time to debug the new SSH failures. It seems the `SaltVirtualEnv` does not respect the pinned requirements, causing `networkx==3.3` to be installed, which is incompatible with the system Python on the failing systems (3.9). There's a bug in the helper code, which can be fixed easily:

```patch
diff --git a/tests/support/helpers.py b/tests/support/helpers.py
index ab435d248f..87374f0e62 100644
--- a/tests/support/helpers.py
+++ b/tests/support/helpers.py
@@ -1673,11 +1673,8 @@ class VirtualEnv:
         kwargs.setdefault("stdout", subprocess.PIPE)
         kwargs.setdefault("stderr", subprocess.PIPE)
         kwargs.setdefault("universal_newlines", True)
-        env = kwargs.pop("env", None)
-        if env:
-            env = self.environ.copy().update(env)
-        else:
-            env = self.environ
+        env = self.environ.copy()
+        env.update(kwargs.pop("env", None) or {})
         proc = subprocess.run(args, check=False, env=env, **kwargs)
         ret = ProcessResult(
             returncode=proc.returncode,
@@ -1780,8 +1777,7 @@ class SaltVirtualEnv(VirtualEnv):
         self.install(RUNTIME_VARS.CODE_DIR)

     def install(self, *args, **kwargs):
-        env = self.environ.copy()
-        env.update(kwargs.pop("env", None) or {})
+        env = kwargs.pop("env", None) or {}
         env["USE_STATIC_REQUIREMENTS"] = "1"
         kwargs["env"] = env
         return super().install(*args, **kwargs)
```

Note: I would suggest pinning `networkx==3.1` for all environments if possible since Salt-SSH often targets older systems and `networkx` will become a hard requirement with these changes. >= 3.3 requires 3.10, >= 3.2 requires 3.9.

twangboy commented:
@bdrx312 You wanna address @lkubb's comments?

bdrx312 commented:
> @bdrx312 You wanna address @lkubb's comments?

Sure. I will take a stab at it now.

bdrx312 commented:
> Note: I would suggest pinning `networkx==3.1` for all environments if possible since Salt-SSH often targets older systems and `networkx` will become a hard requirement with these changes. >= 3.3 requires 3.10, >= 3.2 requires 3.9.

I don't see any other packages in base.txt that set a max version, they all have `>=` or no version. In the constraints.txt the packages have a max set but those are not also in base.txt. Is setting `networkx==3.1` in base.txt the way to do it?

lkubb commented:
> I don't see any other packages in base.txt that set a max version, they all have `>=` or no version. In the constraints.txt the packages have a max set but those are not also in base.txt. Is setting `networkx==3.1` in base.txt the way to do it?

Good question. I would suggest to include `networkx<=3.1` in `requirements/constraints.txt` since - I assume - it only influences the static requirements without hard-capping or outright forcing it for a regular pip install. Not sure if that's the correct way though.

bdrx312 commented:
> > I don't see any other packages in base.txt that set a max version, they all have `>=` or no version. In the constraints.txt the packages have a max set but those are not also in base.txt. Is setting `networkx==3.1` in base.txt the way to do it?
> 
> Good question. I would suggest to include `networkx<=3.1` (or `networkx<3.2`) in `requirements/constraints.txt` since - I assume - it only influences the static requirements without hard-capping or outright forcing it for a regular pip install. Not sure if that's the correct way though.

All the required packages seem like they would have this problem where they need to be pinned to a version supported by the oldest supported version of python. None of them have a max set, yet they all happen to have the same version pinned in the static.txt files. Is this just a coincidence and is a lurking problem waiting to hit? 

```console
$ grep -r -Ee "\b(jinja2|yaml|tornado|msgpack|networkx|certifi|singledispatch|concurrent|singledispatch_helpers|ssl_match_hostname|markupsafe|backports_abc|looseversion|packaging)==" --include=linux.txt -h requirements | sort -u
certifi==2023.07.22
jinja2==3.1.4
looseversion==1.3.0
markupsafe==2.1.3
msgpack==1.0.7
networkx==3.1
packaging==23.1
tornado==6.3.3
```

bdrx312 commented:
```
     def install(self, *args, **kwargs):
-        env = self.environ.copy()
-        env.update(kwargs.pop("env", None) or {})
+        env = kwargs.pop("env", None) or {}
         env["USE_STATIC_REQUIREMENTS"] = "1"
         kwargs["env"] = env
         return super().install(*args, **kwargs)
```
What exactly is the problem here? Why does it need to start with a clean environment instead of the having existing environment variables of the class/process?

lkubb commented:
> All the required packages seem like they would have this problem where they need to be pinned to a version supported by the oldest supported version of python. None of them have a max set, yet they all happen to have the same version pinned in the static.txt files. Is this just a coincidence and is a lurking problem waiting to hit?

Afaik it could become a problem in theory, but the required ones are conservative in their minimum requirements. `networkx` seems pretty aggressive with only >=3.10 being supported by the latest version. It's probably written with a data scientist application in mind.

> What exactly is the problem here?

`env = self.environ.copy().update(env)`

The problem is that `env` is always `None` because `.update` works on the instance itself and does not return anything.

bdrx312 commented:
> > All the required packages seem like they would have this problem where they need to be pinned to a version supported by the oldest supported version of python. None of them have a max set, yet they all happen to have the same version pinned in the static.txt files. Is this just a coincidence and is a lurking problem waiting to hit?
> 
> Afaik it could become a problem in theory, but the required ones are conservative in their minimum requirements. `networkx` seems pretty aggressive with only >=3.10 being supported by the latest version. It's probably written with a data scientist application in mind.
> 
> > What exactly is the problem here?
> 
> `env = self.environ.copy().update(env)`
> 
> The problem is that `env` ~is~ becomes always `None` because `.update` works on the instance itself and does not return anything (thus `USE_STATIC_REQUIREMENTS=1` is absent). The rest is just redundant.

Ah yes, update doesn't return the dict itself. I wasn't even thinking about that.

bdrx312 commented:
> > All the required packages seem like they would have this problem where they need to be pinned to a version supported by the oldest supported version of python. None of them have a max set, yet they all happen to have the same version pinned in the static.txt files. Is this just a coincidence and is a lurking problem waiting to hit?
> 
> Afaik it could become a problem in theory, but the required ones are conservative in their minimum requirements. `networkx` seems pretty aggressive with only >=3.10 being supported by the latest version. It's probably written with a data scientist application in mind.
> 
> > What exactly is the problem here?
> 
> `env = self.environ.copy().update(env)`
> 
> The problem is that `env` ~is~ becomes always `None` because `.update` works on the instance itself and does not return anything (thus `USE_STATIC_REQUIREMENTS=1` is absent). The rest is just redundant.



> > > All the required packages seem like they would have this problem where they need to be pinned to a version supported by the oldest supported version of python. None of them have a max set, yet they all happen to have the same version pinned in the static.txt files. Is this just a coincidence and is a lurking problem waiting to hit?
> > 
> > 
> > Afaik it could become a problem in theory, but the required ones are conservative in their minimum requirements. `networkx` seems pretty aggressive with only >=3.10 being supported by the latest version. It's probably written with a data scientist application in mind.
> > > What exactly is the problem here?
> > 
> > 
> > `env = self.environ.copy().update(env)`
> > The problem is that `env` ~is~ becomes always `None` because `.update` works on the instance itself and does not return anything (thus `USE_STATIC_REQUIREMENTS=1` is absent). The rest is just redundant.
> 
> Ah yes, update doesn't return the dict itself. I wasn't even thinking about that.

What is the problem with the other change though?
```
-        env = self.environ.copy()
-        env.update(kwargs.pop("env", None) or {})
+        env = kwargs.pop("env", None) or {}
```

Shouldn't it start with the existing environment variables and then update them with the additional variables?

lkubb commented:
> What is the problem with the other change though?

It's just unnecessary to do it twice. `SaltVirtualEnv.install` currently

- fetches current env vars
- merges overrides from params
- sets `USE_STATIC_REQUIREMENTS=1`

https://github.com/saltstack/salt/blob/e4c1da4323f6334c0a14e9a68392a0c12a37e117/tests/support/helpers.py#L1782-L1787

This is passed into `VirtualEnv.install`, which delegates to `VirtualEnv.run`, which attempts to

- fetch current env vars
- merge overrides from params (again)
- (then fails to actually associate it to the final variable)

https://github.com/saltstack/salt/blob/e4c1da4323f6334c0a14e9a68392a0c12a37e117/tests/support/helpers.py#L1662-L1663

https://github.com/saltstack/salt/blob/e4c1da4323f6334c0a14e9a68392a0c12a37e117/tests/support/helpers.py#L1670-L1680

Setting env defaults should only be done in the base class, right before the call is executed.

lkubb commented on salt/utils/requisite.py:
```diff
@@ -0,0 +1,470 @@
+from enum import Enum, auto
+from typing import TYPE_CHECKING, Any
+
+import networkx as nx  # pylint: disable=3rd-party-module-not-gated
```
nitpick: I don't think this is necessary anymore since `networkx` was added to Pylint's allowed third party modules.

```suggestion
import networkx as nx
```

lkubb commented on salt/modules/chroot.py:
```diff
@@ -347,7 +347,9 @@ def sls(root, mods, saltenv="base", test=None, exclude=None, **kwargs):
         # Compile and verify the raw chunks
-        chunks = st_.state.compile_high_data(high_data)
+        chunks, errors = st_.state.compile_high_data(high_data)
+        if errors:
+            return errors
```
note: If this is an error condition, it should set the retcode accordingly.

```suggestion
        if errors:
            __context__["retcode"] = salt.defaults.exitcodes.EX_STATE_COMPILER_ERROR
            return errors
```

bdrx312 replied:
Good suggestions/finds. Should this retcode also be set in all the existing `if errors:` lines above too then like on line 339 and 344?

lkubb replied:
Yes. Weird, `salt.defaults.exitcodes` is even imported, but never used in this module.

lkubb reviewed (commented):
Just some minor comments.

Again, thanks a lot for significantly improving the state compiler! I hope this can land soon, especially since it paves the way for some further fixes.

lkubb commented on requirements/base.txt:
```diff
@@ -5,6 +5,8 @@ jmespath
 PyYAML
 MarkupSafe
+# pin to a version available on all supported python versions so salt-ssh can run on older targets
+networkx==3.1
```
The `requirements/static/*/py*/*` files are pinned to exact versions generated using `pip-compile`, which respects the constraints.

lkubb commented on tests/support/helpers.py:
```diff
@@ -1675,7 +1675,8 @@ def run(self, *args, **kwargs):
         kwargs.setdefault("universal_newlines", True)
         env = kwargs.pop("env", None)
         if env:
-            env = self.environ.copy().update(env)
+            env = self.environ.copy()
+            env.update(env)
```
This variant of my fix for the `.update` issue introduces another one since you're reassigning the provided `env` kwarg with the copy of the env var dict and update the latter with itself. `USE_STATIC_REQUIREMENTS` is thus still absent, which is why the tests are still failing.

You could either get rid of the branching since it's relatively unnecessary (like I did) or rename the first `env` from kwargs to something else.

bdrx312 replied:
Odd. I thought I updated this to do it in 2 statements you show here.

bdrx312 replied:
Oh nevermind I see what you are saying. I made the change too quickly I guess without really looking at it.

bdrx312 replied:
Thanks for your help.

jfindlay commented:
Is it possible this can get into 3008.0?

frebib commented:
Is this going to be merged at some point?

bdrx312 committed:
Use importlib.metadata for python >= 3.10

Add slots to DependencyGraph

Pinned networkx to 3.1

Pinned networkx to latest version supported on python 3.8 to support running salt-ssh to systems with python 3.8

Fixed assignment of env in support helpers VirtualEnv

Restrict networkx version with contraints instead of in base

Fixed setting context retcode in some places where errors are returned.
Removed unnecessary pylint ignore comment.
Reverted importlib-metadata version bump.

Added networkx to versions report

Actually fixed VirtualEnv env assignment from kwargs

bdrx312 commented:
> This needs to be rebased and conflicts resolved

I rebased off the latest master and resolved the conflicts.

twangboy commented on tests/pytests/unit/state/test_state_compiler.py:
```diff
@@ -855,128 +862,6 @@ def test_call_chunk_sub_state_run(minion_opts):
             assert sub_state["__sls__"] == "external"
 
 
-def test_aggregate_requisites(minion_opts):
```
Is there a reason why this test is removed?

bdrx312 replied:
If I am remembering correctly (since I removed this half a year ago), because the test is not correct and is not how the ordering and mod aggregate works. If I am remembering correctly the requisites of the states stay the same and the network is just updated to determine the requisite ordering.

twangboy commented on tests/pytests/unit/state/test_state_compiler.py:
```diff
@@ -1064,66 +960,11 @@ def test_mod_aggregate(minion_opts):
-            assert low_ret["pkgs"] == ["figlet", "sl"]
-
-
-def test_mod_aggregate_order(minion_opts):
```
And this one?

bdrx312 replied:
This one is essentially a duplicate test and also is not correct.

bdrx312 replied:
If I am remembering correctly (since I removed this half a year ago), because the test is not correct and is not how the ordering and mod aggregate works. If I am remembering correctly the requisites of the states stay the same and the network is just updated to determine the requisite ordering.
